### ISSUE TITLE
Super+number has opposite behavior than expected when it comes to workspaces

### ISSUE THREAD

**Issue by Asghwor** [linuxmint/cinnamon#10668]
```
 * Cinnamon 5.2.7
 * Distribution Mint 20.3
 * Geforce 1660 Mobile &  nvidia-driver-470
 * 64 bit
 ```

**Issue**

1. When pressing super+number to open application that's already open in another workspace, I'm dragged to that workspace instead of a new instance of the application opening.
2. When pressing super+number to cycle through the windows of an application that's not open in another workspace, or focus the single one that's open, it creates a new window instead.
3. When pressing super+number to cycle through the windows of an application that's open in the current workspace and another workspace, it drags me to that other workspace instead of cycling through the windows of the current workspace.

This takes a while to start happening and the third case is rarer.

**Steps to reproduce**

For the first case:
1. Use Cinnamon for some time.
2. Open a few windows here and there.
3. Go to a workspace when a certain windows **is not open**.
4. Press super+number, where "number" is its position on the taskbar.
5. You will be dragged to another workspace where the window is open.

For the second case:
1. Use Cinnamon for some time.
2. Open a few windows here and there.
3. Go to a workspace when a certain windows **is open**.
4. Press super+number, where "number" is its position on the taskbar.
5. A new window will be open instead of cycling through the windows that were already open.

For the third case (rarer):
1. Use Cinnamon for some time.
2. Open a few windows here and there.
3. Open windows for the same application in multiple workspaces.
4. Go to a workspace when a certain windows **is open**.
5. Press super+number, where "number" is its position on the taskbar.
6. You'll be dragged to another workspace where the window is open instead of cycling through the windows that were already open.

**Expected behaviour**

1. A new window would be open in the current workspace.
2. You would cycle through the windows that are open in this workspace.
3. You would cycle through the windows that are open in this workspace.

**Other information**

1. This was already happening on Mint 20.2.
2. Restarting Cinnamon can make it work correctly for a few seconds, sometimes.

[xsession-errors.txt](https://github.com/linuxmint/cinnamon/files/8057801/xsession-errors.txt)

**Comment by Asghwor** [comment]
Still a problem in Cinnamon 5.4.9 in Linux Mint 21.

**Comment by Asghwor** [comment]
I've noticed the two incorrect behaviors, creating new windows in the same workspace and dragging the user to a workspace where the window already exists, happen separately now.

More specifically, being dragged to another workspace where the app is already open when pressing Super+[number], happens at the start. After a while, pressing Super+[number] starts spawning new windows no matter how many are already open in the current workspace.

The following are videos of both cases:

[infinite-spawning.webm](https://user-images.githubusercontent.com/30949004/184494934-5654ca0b-3cc0-46b1-b641-e8c8c17f97c7.webm)

[dragged-to-another-workspace.webm](https://user-images.githubusercontent.com/30949004/184494937-d21020a3-f901-4579-9af0-4626ab798e71.webm)

**Comment by Asghwor** [comment]
Posting xsession-errors here since it fails to do so in the main post.

[xsession-errors.txt](https://github.com/linuxmint/cinnamon/files/9331515/xsession-errors.txt)

**Comment by DirkHaar** [comment]
I'm not sure if I deleted the standard key binding - but Super+n (for N=1 to 9) does nothing for me.
You might have defined your own key bindings in 20.2.

**Comment by Asghwor** [comment]
I haven't really ever set the Super+[number] shortcuts. I've also installed Mint 21 from scratch, so it couldn't have been a forgotten dotfile or something.

**Comment by DirkHaar** [comment]
ALT + number opens the corresponding TAB in the currently opened Firefox session.

I'm quite sure you either set up your own ke bindings in the past (by some add on maybe?) or you mix something up.

**Comment by Asghwor** [comment]
No, I haven't. As I said, these keybindings come with the system.

**Comment by DirkHaar** [comment]
Can you show the corresponding section in the key binding dialog (menu/keyboard/...)? 
In my version there is no such setting; there is
naigation:workspaces: go to workspace n (-12)
and
windows:move windows(*):move window to workspace n (1-12)

Maybe you had this in another DE (KDE) before?

(*forgive me, my version is German so the titles may differ)

**Comment by Asghwor** [comment]
I just installed Mint on a formatted partition. Perhaps you had another DE before and that's why it's behaving differently from stock Mint?

**Comment by DirkHaar** [comment]
No, I run Mint Cinnamon for more than ten years.

**Comment by DirkHaar** [comment]
Can you add a screenshot of your keyboard settings?
Go to tab "key bindings", select "Workspaces", and then "direct navigation" on the left.

**Comment by Asghwor** [comment]
Perhaps you've been recycling your `/home` and that's been causing your issue of not having the appropriate shortcuts.

Anyway, I've taken a look at the source code and there are functions related to Super+Number shortcuts in there. I think it might be getting confused when looking at the current open windows in the workspaces for some reason.

**Comment by Asghwor** [comment]
I tried to debug this issue, and the weird thing is that, on the `AppGroup.onAppKeyPress` method, the `AppGroup.groupState.metaWindows` is always empty but, on `AppGroup.calcWindowNumber`, the array is properly filled.

It feels like some memoization problem, but I can't tell where.
